File: GFG/Stacks/stockspan.py
import ds

def sspan(arr):
    n=len(arr)
    st1=ds.Stack()
    st2=ds.Stack()
    st1.push(arr[0])
    print("1",end=" ")
    i=1
    while(i<n):
        q=1
        if(arr[i]<=st1.top()):
            st1.push(arr[i])
            print(q,end=" ")
            i+=1
        else:
            while(st1.size()>0 and arr[i]>st1.top() ):
                q+=1
                st2.push(st1.pop())
            while(st2.size()>0):
                st1.push(st2.pop())
            st1.push(arr[i])
            print(q,end=" ")
            i+=1
arr=[13,15,12,14,16,8,6,4,10,30]
sspan(arr)

        

File: GFG/Stacks/stockspan.cpp
#include<iostream>
#include"../ds.h"
using namespace std;

void span(vector<int> arr){
    Stack<int> st1,st2;
    int n=arr.size();
    int i=1;
    st1.push(arr[0]);
    cout<<"1 ";
    while(i<n){
        int q=1;
        if(arr[i]<=st1.peek()){
            st1.push(arr[i]);
            cout<<q<<" ";
            i++;
        }
        else{
            if(arr[i]>st1.peek()){
                while(st1.size()>0 && st1.peek()<=arr[i]   ){
                    st2.push(st1.pop());
                    q++;
                }
                
                while(st2.size()>0){
                    st1.push(st2.pop());
                }
                st1.push(arr[i]);
                cout<<q<<" ";
            }
            i++;
        }
    }
    return ;
}

int main(){
    // vector<int> arr {13,15,12,14,16,8,6,4,10,30};
    vector<int> arr {10,20,30,40};
    span(arr);
    return 0;
}